File: gfdesignpatterns/Ruby/Builder.rb
class Show 
	def initialize()
		@name
		@type
		@year
	end
	
	def name=(newName)
		@name = newName
	end
	
	def type=(newType)
		@type = newType
	end
	
	def year=(newYear)
		@year = newYear
	end
	
	def to_s
		"Name is: #{@name}  The Year is: #{@year}  The Type is: #{@type}"
	end
end

class Builder 
	def initialize()
		@show1 = Show.new();
	end
	
	def buildname(newName)
		@show1.name=newName
	end
	
	def buildyear(newYear)
		@show1.year=newYear
	end
	
	def buildtype()
		raise NotImplementedError, "#{self.class.type} does not implement buildType()"
	end
	
	def to_s
		@show1.to_s
	end
end

class BuildShort < Builder
	def buildtype()
		@show1.type="Short"
	end
end

class BuildMovie < Builder
	def buildtype()
		@show1.type="Movie"
	end
end

class Director
	def initialize()
	end
	
	def construct(builder, name, year)
		builder.buildname(name)
		builder.buildtype()
		builder.buildyear(year)
	end
end

director1 = Director.new()
movieBuilder = BuildMovie.new()
shortBuilder = BuildShort.new()

director1.construct(movieBuilder,"Inside Out",2015)
show1 = movieBuilder.to_s

director1.construct(movieBuilder,"Finding Nemo",2003)
show2 = movieBuilder.to_s

director1.construct(shortBuilder,"One Man Band",2005)
show3 = shortBuilder.to_s

director1.construct(shortBuilder,"Presto",2008)
show4 = shortBuilder.to_s

puts "Show 1"
puts show1.to_s

puts "Show 2"
puts show2.to_s

puts "Show 3"
puts show3.to_s

puts "Show 4"
puts show4.to_s


	
	
	
	
	
	
	
	
	
	
	

File: gfdesignpatterns/gfdesignpatterns/builder/builder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace gfdesignpatterns.builder
{

        public class Director
        {
            public void construct(Builder builder, string name, string year, string mainChar)
            {
                builder.buildName(name);
                builder.buildType();
                builder.buildYear(year);
                builder.buildMainChar(mainChar);
            }
        }

        public abstract class Builder
        {
            protected Show _show = new Show();
            public abstract void buildType();
            public void buildName(string name)
            {
                _show.name = name;
            }

            public void buildYear(string year)
            {
                _show.year = year;
            }


            public void buildMainChar(string mainChar)
            {
                _show.mainChar = mainChar;
            }
            public string getResult()
            {
                return _show.print();
            }
        }

        class buildMovie : Builder
        {
            public override void buildType()
            {
                this._show.type = "Movie";
            }

        }

        class buildShort : Builder
        {
            public override void buildType()
            {
                this._show.type = "Short";
            }
        }


        public class Show
        {
            private string _name = "";
            private string _year = "";
            private string _mainChar = "";
            private string _type = "";

            public string name
            {
                set { _name = value; }
            }
            public string year
            {
                set { _year = value; }
            }
            public string mainChar
            {
                set { _mainChar = value; }
            }
            public string type
            {
                set { _type = value; }
            }
            public string print ()
            {
                return string.Format("The name of this show is {0}, <br> It is a {1},<br> made in {2},<br> and the main character is {3}  ",_name,_type,_year,_mainChar);
                
            }
        }

    }
